File: Solutions/Python/adding_spaces_to_a_string.py
"""
    [MEDIUM]
    2109. Adding Spaces to a String

    Concepts:
    - string
    - two pointers

    Stats:
        Runtime | 131 ms    [Beats 31.78%]
        Memory  | 48.7 MB   [Beats 58.55%]
"""

class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:

        n = len(s)
        k = len(spaces)
        res = [""]*(n + k)

        i = 0   # spaces pointer
        j = 0   # s pointer

        while j < n:
            # if has spaces to insert and reached space index...
            if i < k and j == spaces[i]:
                res[j + i] = " "
                i += 1

            # if no spaces to insert or not at space index...
            else:
                res[j + i] = s[j]
                j += 1

        return "".join(res)

File: Solutions/Cpp/adding_spaces_to_a_string.cpp
/*
    [MEDIUM]
    2109. Adding Spaces to a String

    Concepts:
    - string
    - two pointers

    Stats:
        Runtime | 19 ms     [Beats 63.55%]
        Memory  | 85.42 MB  [Beats 19.63%]
*/

class Solution {
public:
    string addSpaces(string s, vector<int>& spaces) {
        int n = s.size();
        int k = spaces.size();

        int i = 0;
        int j = 0;

        string res;

        while (j < n) {
            if (i < k && j == spaces.at(i)) {
                // string& append(const char* c)
                res.append(" ");
                i++;
            } else {
                // void push_back (char c)
                res.push_back(s.at(j));
                j++;
            }
        }

        return res;
    }
};